bundles-lib/src/main/java/org/apache/metron/bundles/BundleClassLoaders.java

```diff
@@ -0,0 +1,353 @@
+
+        // load the bundle details which includes bundle dependencies
+        for (final FileObject bundleFile : bundleDirContents) {
+          if(!bundleFile.exists() || !bundleFile.isFile()) {
+            continue;
+          }
+          BundleDetails bundleDetail = null;
+          try {
+            bundleDetail = getBundleDetails(bundleFile, props);
+          } catch (IllegalStateException e) {
+            logger.warn("Unable to load BUNDLE {} due to {}, skipping...",
+                new Object[]{bundleFile.getURL(), e.getMessage()});
+          }
+
+          // prevent the application from starting when there are two BUNDLEs with same group, id, and version
+          final String bundleCoordinate = bundleDetail.getCoordinate().getCoordinate();
+          if (bundleCoordinatesToBundleFile.containsKey(bundleCoordinate)) {
+            final String existingBundleWorkingDir = bundleCoordinatesToBundleFile
+                .get(bundleCoordinate);
+            throw new IllegalStateException(
+                "Unable to load BUNDLE with coordinates " + bundleCoordinate
+                    + " and working directory " + bundleDetail.getBundleFile()
+                    + " because another BUNDLE with the same coordinates already exists at "
+                    + existingBundleWorkingDir);
+          }
+
+          bundleDetails.add(bundleDetail);
+          bundleCoordinatesToBundleFile.put(bundleCoordinate,
+              bundleDetail.getBundleFile().getURL().toURI().toString());
+        }
+
+        // attempt to locate the jetty bundle
+        ClassLoader jettyClassLoader = null;
+        for (final Iterator<BundleDetails> bundleDetailsIter = bundleDetails.iterator();
+            bundleDetailsIter.hasNext(); ) {
+          final BundleDetails bundleDetail = bundleDetailsIter.next();
+          // populate bundle lookup
+          idBundleLookup.computeIfAbsent(bundleDetail.getCoordinate().getId(),
+              id -> new HashSet<>()).add(bundleDetail.getCoordinate());
+        }
+
+        int bundleCount;
+        do {
+          // record the number of bundles to be loaded
+          bundleCount = bundleDetails.size();
+
+          // attempt to create each bundle class loader
+          for (final Iterator<BundleDetails> bundleDetailsIter = bundleDetails.iterator();
+              bundleDetailsIter.hasNext(); ) {
+            final BundleDetails bundleDetail = bundleDetailsIter.next();
+            final BundleCoordinate bundleDependencyCoordinate = bundleDetail
+                .getDependencyCoordinate();
+
+            // see if this class loader is eligible for loading
+            ClassLoader potentialBundleClassLoader = null;
+            if (bundleDependencyCoordinate == null) {
+              potentialBundleClassLoader = createBundleClassLoader(fileSystemManager,
+                  bundleDetail.getBundleFile(), ClassLoader.getSystemClassLoader());
```

Comment:
It depends if the parent classloader has what is required, and any of the 'extraneous' bundles are accessed and they are not satisfied with the current parent.

Our bundle host is currently the metron-parser shaded jar.  Any of our shaded jars ( topologies ) should be fine for this, given that they would only try to load extensions of types they know about.

NiFi seems to want every bundle to have access to a Jetty Server.